File: DS/Heap.py
import heapq

# --- TEMPLATE START ---

# Min-Heap Functions
# heapq implements a min-heap by default.
# The heap is just a standard list.
def min_heap_push(h, item):
    heapq.heappush(h, item)

def min_heap_pop(h):
    return heapq.heappop(h)

# Max-Heap Functions (by storing negative values)
# This works for numbers. For tuples, negate the priority element.
def max_heap_push(h, item):
    heapq.heappush(h, -item)

def max_heap_pop(h):
    return -heapq.heappop(h)

def max_heap_top(h):
    return -h[0]

# --- TEMPLATE END ---

def solve():
    # --- Example 1: Min-Heap of integers ---
    min_h = []
    min_heap_push(min_h, 20)
    min_heap_push(min_h, 5)
    min_heap_push(min_h, 10)
    print(f"Min-Heap top: {min_h[0]}") # 5
    min_heap_pop(min_h)

    # --- Example 2: Max-Heap of integers ---
    max_h = []
    max_heap_push(max_h, 20)
    max_heap_push(max_h, 5)
    max_heap_push(max_h, 100)
    print(f"Max-Heap top: {max_heap_top(max_h)}") # 100
    max_heap_pop(max_h)

    # --- Example 3: Min-Heap of tuples (useful for Dijkstra) ---
    # Tuples are compared element by element, which is perfect.
    dijkstra_pq = []
    min_heap_push(dijkstra_pq, (10, 1)) # (distance, node)
    min_heap_push(dijkstra_pq, (5, 2))
    min_heap_push(dijkstra_pq, (10, 0))
    print(f"Dijkstra next node: {dijkstra_pq[0]}") # (5, 2)

    # --- Example 4: Max-Heap of custom data using tuples ---
    # To get max-heap behavior, negate the priority element of the tuple.
    item_max_h = []
    # Push as (-priority, data1, data2, ...)
    min_heap_push(item_max_h, (-50, 1, "Item A"))
    min_heap_push(item_max_h, (-100, 2, "Item B"))
    
    top_item = item_max_h[0]
    print(f"Max value item ID: {top_item[1]}") # 2
    print(f"Max value: {-top_item[0]}") # 100

solve()

File: DS/Heap.cpp
#include<bits/stdc++.h>
#define ll long long
using namespace std;

// --- TEMPLATE START ---

// Use these aliases for speed and clarity in contests
template<typename T>
using MinHeap = std::priority_queue<T, std::vector<T>, std::greater<T>>;

template<typename T>
using MaxHeap = std::priority_queue<T>;

// For custom structs, you need to define a comparison
struct Item {
    int id;
    int value;

    // For MaxHeap: a < b means 'a' has lower priority than 'b'
    bool operator<(const Item& other) const {
        return this->value < other.value;
    }

    // For MinHeap: a > b means 'a' has lower priority than 'b'
    // You only need one of these depending on the heap type.
    // An alternative to overloading '>' is to create a custom comparator struct.
    bool operator>(const Item& other) const {
        return this->value > other.value;
    }
};

template<typename T>
struct mincompare {
    bool operator()(const T& a, const T& b) const {
        return a > b; // For MinHeap
    }
};
template<typename T>
struct maxcompare {
    bool operator()(const T& a, const T& b) const {
        return a < b; // For MaxHeap
    }
};

// --- TEMPLATE END ---

void solve() {
    // --- Example 1: Min-Heap of integers ---
    MinHeap<int> minH;
    minH.push(20);
    minH.push(5);
    minH.push(10);
    std::cout << "Min-Heap top: " << minH.top() << std::endl; // 5
    minH.pop();

    // --- Example 2: Max-Heap of integers ---
    MaxHeap<int> maxH;
    maxH.push(20);
    maxH.push(5);
    maxH.push(100);
    std::cout << "Max-Heap top: " << maxH.top() << std::endl; // 100
    maxH.pop();

    // --- Example 3: Min-Heap of pairs (useful for Dijkstra) ---
    // Pairs are compared lexicographically (by first element, then second).
    // This is exactly what you want for (distance, node) pairs.
    MinHeap<std::pair<int, int>> dijkstra_pq;
    dijkstra_pq.push({10, 1}); // {distance, node}
    dijkstra_pq.push({5, 2});
    dijkstra_pq.push({10, 0});
    std::cout << "Dijkstra next node: (" << dijkstra_pq.top().first << ", " 
              << dijkstra_pq.top().second << ")" << std::endl; // (5, 2)

    // --- Example 4: Max-Heap of custom structs ---
    MaxHeap<Item> itemMaxH;
    itemMaxH.push({1, 50});
    itemMaxH.push({2, 100});
    std::cout << "Max value item ID: " << itemMaxH.top().id << std::endl; // 2

    // extract k numbers from a priority queue
    int k = 2;
    std::cout << "Extracting top " << k << " from Max-Heap:" << std::endl;
    while (k-- && !itemMaxH.empty()) {
        std::cout << "Item ID: " << itemMaxH.top().id << ", Value: " << itemMaxH.top().value << std::endl;
        itemMaxH.pop();
    }

}

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);
    solve();
    return 0;
}